sip/science_pipeline_workflows/ingest_visibilities/recv_c/test/Tests.cpp

```diff
@@ -107,3 +108,115 @@ TEST(Stream, test_stream_decode)
     receiver_free(receiver);
     free(buf);
 }
+
+
+TEST(Stream, test_stream_receive)
+{
+    /* Verify stream_decode() function and verify if correct
+    data is received */
+    typedef enum _item_ids {
+        heap_counter=0x01,
+        heap_size,
+        heap_offset,
+        packet_payload_length,
+        visibility_baseline_count=0x6005,
+        scan_ID=0x6008,
+        visibility_data=0x600A
+    } item_id;
+
+    typedef struct _item_block {
+        item_id id;
+        uint64_t val;
+        uint64_t imm;
+    } item_block;
+
+    int num_baselines = 512 *513 / 2;
+    struct  DataType vis_data = {0,0,{{1.5,2.5},{3.5,4.5},{5.5,6.5},{7.5,8.5}}};
+    // float vis_data[2];
+    // vis_data[0] = (float)rand()/(float)(RAND_MAX/10.);
+    // vis_data[1] = (float)rand()/(float)(RAND_MAX/10.);
+    uint64_t scan_id=0, *scan_ptr=NULL;
+    void *heap_start;
+    struct DataType * vis_ptr;
+
+    const item_block p_items[] = { 
+	{heap_counter, 2, 1}, 
+	{heap_size, 16, 1},
+	{heap_offset, 0, 1}, 
+	{packet_payload_length, 16, 1},
+	{visibility_baseline_count, (uint64_t) num_baselines, 1},
+	{scan_ID, 0, 0},
+	{visibility_data, 8, 0}
+    };
+    
+    const unsigned int num_items = sizeof(p_items) / sizeof(item_block);
+
+    // Allocate memory for a SPEAD packet.
+    unsigned char* buf = (unsigned char*) malloc(8 + num_items * 8 + 8 + sizeof(vis_data));
+    ASSERT_TRUE(buf != NULL);
+
+    // Define SPEAD flavour.
+    const unsigned char heap_address_bits = 48;
+    const unsigned char item_id_bits = 64 - heap_address_bits - 1;
+
+    // Construct SPEAD packet header.
+    buf[0] = 'S';
+    buf[1] = 4;
+    buf[2] = (item_id_bits + 1) / 8;
+    buf[3] = heap_address_bits / 8;
+    buf[4] = 0;
+    buf[5] = 0;
+    buf[6] = 0;
+    buf[7] = (unsigned char) num_items; 
+
+    // Construct SPEAD packet payload containing items in
+    // "immediate addressing mode".
+    uint64_t* items = (uint64_t*) &buf[8];
+    uint64_t heap_count;
+    for (unsigned int i = 0; i < num_items; i++)
+    {
+        const uint64_t item_id = (uint64_t) p_items[i].id;
+        const uint64_t item_addr_or_val = (uint64_t) p_items[i].val;
+	    const uint64_t imm_or_abs = p_items[i].imm << 63;
+        //printf("Item Value: %i\n", item_val[i]);
+        const uint64_t item =
+                imm_or_abs |
+                (item_id << heap_address_bits) |
+                item_addr_or_val;
+        items[i] = htobe64(item);
+	switch (p_items[i].id)
+	{
+	    case heap_counter : heap_count = p_items[i].val; break;
+	}
+    }
+
+    // get start of heap
+    heap_start = buf + 8 + num_items * sizeof(uint64_t);
+
+    // insert scan_id at heap_start, 8 bytes
+    scan_ptr = (uint64_t *) heap_start;
+    *scan_ptr = scan_id;
+    *scan_ptr = htobe64(scan_id);
+    ++scan_ptr;
+    vis_ptr = (struct DataType *) scan_ptr;
```

Comment:
I think it would be a bit clearer to write this instead, and remove the line above it:
```suggestion
    struct DataType* vis_ptr = (struct DataType*) (scan_ptr + 1);
```
The separate declaration of `vis_ptr` on line 140 is then also not required.

Reply:
I have to admit a preference to putting all variable declarations at the start of a code block. We may be C programmers, but that doesn't make us anarchists :-)

Reply:
My motivation for flagging this was because it's logically wrong to increment a variable called `scan_ptr`, which then makes it no longer point to the scan ID. Code like this is unsafe, as any future use of `scan_ptr` would then be an error.